 #H12
 Harjoituksen tarkoituksena on aiheuttaa ja tutkia onglemia Djangon tuotantotyyppisessä asennuksessa.
 
### Laitteisto
 
* Käyttöjärjestelmä	Microsoft Windows 10 Enterprise LTSC 64 bit
* Prosessori i5-6600
* RAM 16 GB
* Virtuaaliohjelmisto : Oracle VM VirtualBox
* Virtuaalikoneen käyttöjärjestelmä: Debian-live-11.6.0-amd64-xfce+nonfree.iso
* Virtuaalipalvelin: Linode Nanode 1 GB





## Alustus (11:54)

Aloitin harjoituksen tarkistamalla, että Djangon tuotantoasennus toimii odetetulla tavalla.

    
Aloitin siirtymällä viime tehtävässä luotuun publicswgi kansioon ja ja käynnistämällä virtuaaliympäristön komennoilla:


    $ cd publicswgi
    $ source env/bin/activate
    
Tarkistin seuraavaksi Apachen statuksen komennolla:

    $ sudo systemctl apache2
    
 ![image](https://user-images.githubusercontent.com/106889187/223078157-0d051dde-5454-4e3a-8322-5e49781e76e4.png)

 Apache oli käynnissä. Tein myös konfiguraatiotestin komenolla:
 
 
    $ /sbin/apache2ctl configtest
  
  
  ![image](https://user-images.githubusercontent.com/106889187/223078771-a95362df-de21-413b-b2fc-72ffc1ff7dbc.png)
  
  Syntaksi oli OK.

    
  Seuraavaksi kävin selaimessa osoitteissa:
  
      localhost
      localhost/admin
      
  Localhost-sivustolla ei olla luotu sisältöä, niin se näyttää 'Not Found'-ilmoitusta.
  
  ![image](https://user-images.githubusercontent.com/106889187/223081647-0c97b163-8c5b-4390-a6a8-73ec16da30b3.png)

      
 Localhost/admin-sivusto latautui onnistuneesti.
  
  ![image](https://user-images.githubusercontent.com/106889187/223079493-04c078cb-422a-4ae2-9135-72b22872492c.png)
  
  Tarkistin myös, että sivusto pyöri Apachella komennolla:
  
     $ curl -sI localhost|grep Server
     
  ![image](https://user-images.githubusercontent.com/106889187/223079854-745a0a9c-ea28-4fc0-b9b5-af5ca7b28f2e.png)
  
 Kävin katsomassa myös Apachen error-lokeja ja niistä ei löytynyt mitään vikaa.
 
 
  
![image](https://user-images.githubusercontent.com/106889187/223080794-2a797e90-87e7-4954-bf2c-c39c8756f5a9.png)


## Kirjoitusvirhe Python-tiedostossa (12:12)
Luodaan kirjoitusvirhe Python-tiedostoon ja tehdään vianmääritys.

Aloitin menemällä muokkaamaan 'settings.py'-tiedostoa komennolla:

    $ micro /rojekt/rojekt/settings.py
    
Ja muutin rivillä 13 'import' -> 'imort'

![image](https://user-images.githubusercontent.com/106889187/223084181-ba0c5721-6bcc-4047-b12a-ed00ed393cca.png)

Tein Apachen kofniguraatiotestin ja sain syntaksi OK.

 ![image](https://user-images.githubusercontent.com/106889187/223084905-cd09e8ba-831a-4e25-8014-f3e532639053.png)

Seuraavaksi käynnistin Apachen uudestaan komennolla:

    $ sudo systemctl restart apache2
    
Selaimessa osoitteessa 'localhost' sain uuden virheviestin 'Internal Server Error'. Kävin tutkimassa Apachen virhelokeja komennolla:

    $ sudo tail -F /var/log/apache2/error.log 


![image](https://user-images.githubusercontent.com/106889187/223086216-b1bfe4cf-ca60-4844-abc0-f6b9e81052e2.png)

 Viimeisien lokien alussa näkyy päivämäärä ja aika oikeassa muodossa ja sen jälkeen viesti 'wsgi:error', mikä viittaa wsgi-moduulissa olevaan virheeseen. Tämän jälkeen on luoteltu prosessi- ja säie-id. Sitten on listattu, mistä on lähteestä on otettu yhteyttä. Itse viesteissä listataan missä tiedostossa virhe sijaitsee ja mikä virhe on kyseessä (syntaksi error).
 
Kävin muokkaamassa tiedostoa 'settings.py' polussa /home/r01p4r/publicwsgi/rojekt/rojekt/settings.py ja muutin 'imort' -> 'import'.

Tämän jälkeen käynnistin Apachen uudestaan komennolla

    $ sudo systemctl restart apache2
    
  Ja localhost-sivusto toimi taas odotetusti.
  
  ![image](https://user-images.githubusercontent.com/106889187/223088688-bd5022fd-24c0-463b-bba2-670b9b7da644.png)



## Projektikansio väärässä paikassa (12:48)
Luodaan uusi kansio ja siirretään Djangon projekti siihen. Tutkitaan ongelmaa.

Loin ensin uuden kansion publicswgi-hakemistoon komennolla:

    $ mkdir other
    
  ![image](https://user-images.githubusercontent.com/106889187/223090050-754a03b6-ce78-4df5-967b-8f819e91035a.png)
  
  
  Siirsin kansion 'rojekt' uuteen kansioon 'other' komennolla:
  
    $ sudo mv /home/r01p4r/publicwsgi/rojekt /home/r01p4r/publicwsgi/other/
    
    
  ![image](https://user-images.githubusercontent.com/106889187/223096628-9f4ab3c8-a5be-4f54-8bf3-42ff4613c712.png)

  Tein Apachen konfiguraatiotestin ja käynnistin sen uudestaan komennoilla:
  
    $ /sbin/apache2ctl configtest
    $ sudo systemctl restart apache2
  
  
  ![image](https://user-images.githubusercontent.com/106889187/223097096-26b89aa8-5758-4c27-b285-d4cea3663518.png)

Konfiguraatiotesti OK.

Selaimessa localhost antaa nyt '403 Forbidden'-viestiä.

![image](https://user-images.githubusercontent.com/106889187/223097300-62fb170f-32fe-4f9d-a337-82ee27ccb576.png)


Katsoin Apachen virhelokeja seuraavaksi.

![image](https://user-images.githubusercontent.com/106889187/223097811-c013cec1-ab64-4063-abd2-356cc355efaa.png)

Virhelokien perusteella vika on serveri-konfiguraatiossa. Hakemistossa /home/r01p4r/publicswgi/rojekt

Yritin mennä hakemistoon, mutta sain viestin, että kyseistä hakemistoa ei ole. Siirryin seuraavaksi 'publicswgi' hakemistoon ja listasin sen tiedostot komennolla:

    $ ls
    
![image](https://user-images.githubusercontent.com/106889187/223099131-2a94e60e-a5a5-41be-939e-f92f6d807d61.png)

Avasin kansion 'other' ja sen sisältä paljastui haluttu 'rojekt'-kansio. Siirsin kansion oikeaan paikaan komennolla:

    $ sudo mv /home/r01p4r/publicwsgi/other /home/r01p4r/publicwsgi/rojekt/
    
  ![image](https://user-images.githubusercontent.com/106889187/223099705-aac1fa11-f112-400e-8aa9-46022a2f649b.png)

Käynnistin Apachen uudestaan ja sivusto 'localhost' näytti nyt oikeaa viestiä.

## Projektikansiolla väärät oikeudet (13:38)
Muutetaan projektikansiolle väärät oikeudet ja selvitetään vika.

Aloitin ajamalla ohjeessa mainitun komennon:

    $ chmod ugo-rwx rojekt/

Tein apachen konfiguraatiotestin, mikä näytti OK ja käynnistin apachen uudestaan komennolla:

    $ sudo systemctl restart apache2

Ja sain localhostissa virheviestin '403 Forbidden'.

![image](https://user-images.githubusercontent.com/106889187/223101334-c5da2f90-90f5-4497-9727-4a744b7c8689.png)

Apachen virheloki näytti  lauseen 'because search permissions are missing on a component of the pathv' perusteella valittavan puuttuvista oikeuksista polussa '/home/r01p4r/publicwsgi/rojekt/rojekt/wsgi.py').

![image](https://user-images.githubusercontent.com/106889187/223101978-0aef2b21-b271-4ce4-80fb-5a01de10120e.png)

Palautin oikeudet projektille komennoilla:

     $ cd publicwsgi/rojekt/  
     
   ja
   
     $ chmod u+rx rojekt/
     
     
    
 ![image](https://user-images.githubusercontent.com/106889187/223102726-63edbcb7-fcc5-4791-8b1e-d593fa8a3b61.png)
 
 Käynnistin Apachen uudestaan ja sivusto 'localhost' toimi taas odotetusti.
 
 
 
 ## Kirjoitusvirhe Apachen asetustiedostossa (14:04)
 Luodaan kirjoitusvirhe Apachen asetustiedostoon '/etc/apache2/sites-available/rojekt.conf' ja tehdään vianmääritys.
 
 Aloitin muuttamalla 'rojekt.conf'-tiedostoa komennolla
 
    $ sudoedit /etc/apache2/sites-available/rojekt.conf
    
 Ja lisäsin siihen kirjoitusvirheen riville 3: Define TUSER 'r01p4r' -> 'r01p4rds'.
 
 ![image](https://user-images.githubusercontent.com/106889187/223106270-65adf429-9e65-4a84-87fe-638c87b89ef5.png)

 Apachen konfiguraatiotesti osoitti heti virheen 'AH00543: apache2: bad user name r01p4rds' Mikä kertoo, että käyttäjän kohdalla on virhe.
 
 ![image](https://user-images.githubusercontent.com/106889187/223106534-d2162789-50e7-478e-a965-15ed1ed79446.png)
 
 Kokeilin vielä käynnistää Apachen uudestaan, mutta sain viestin:
 
 ![image](https://user-images.githubusercontent.com/106889187/223107056-b851ba5c-5bab-4776-8762-6b885ebf5ea5.png)
 
 Muutin käyttäjän takaisin oikeaksi ja konfiguraatiotesti, sekä 'localhost'-sivu toimi taas odotetusti.
 
 ![image](https://user-images.githubusercontent.com/106889187/223112990-0014b5f9-f1ee-4504-ab92-d16353928056.png)




 
 ## WSGI-moduulin puuttuminen (14:15)
 Poistetaan WSGI-moduuli ja tehdään vianmääritys.
 
 Aloitin poistamalla WSGI-moduulin komennolla:
 
    $ sudo apt-get purge libapache2-mod-wsgi-py3

![image](https://user-images.githubusercontent.com/106889187/223108356-fc5ca087-d6d0-43d7-b5f4-6213172cc13c.png)

Poiston yhteydessä ilmoitetaan, että Apachen konfiguraatio ei toimi ja WSGI-demoni puuttuu tai on väärin ilmoitettu. Suoritin Apachen konfiguraatiotestin ja se osoitti samat virheet.

![image](https://user-images.githubusercontent.com/106889187/223108802-cc76174a-32fd-46a4-aa1f-32185f93cea2.png)

Seuraavaksi yritin käynnistää Apachen vielä uudestaan, mutta sain virheilmoituksen.

![image](https://user-images.githubusercontent.com/106889187/223108971-1b1274da-c9d4-4be5-ad01-ae0b137abdb8.png)

Error-loki osoitti, että WSGI skripti ei löydy tai se ei pysty käynnistymään, jonka seurauksesta Apache2 sulkeutuu.

![image](https://user-images.githubusercontent.com/106889187/223109222-64aafacf-c5c3-4c91-a122-8596998ea9d9.png)

Palautin WSGI-moduulin komennolla:

    $ sudo apt-get install libapache2-mod-wsgi-py3
    
Ja Apachen konfiguraatiotesti, serveri  ja nettisivu toimivat taas odotetusti.

![image](https://user-images.githubusercontent.com/106889187/223110021-dd0a5ecb-7491-4183-83f1-51d91dbddbec.png)

![image](https://user-images.githubusercontent.com/106889187/223113079-7509c892-fe67-48c9-9ef5-ce1336f094e0.png)


 
 ## Lopuksi 
 
 Tässä harjoituksessa tein Djangon tuotantoasennuksen vianmääritystä.
 
 
## Lähteet

Linux Palvelimet 2023 alkukevät, Tero Karvinen (https://terokarvinen.com/2023/linux-palvelimet-2023-alkukevat/)

Deploy Django 4 - Production Install Tero Karvinen - 13.2.2022 (https://terokarvinen.com/2022/deploy-django/)





#### Tehnyt Roi Partanen (2.3.2023)
